src/wallet/api/wallet.cpp

```diff
@@ -1429,75 +1428,83 @@ PendingTransaction *WalletImpl::createTransaction(const string &dst_addr, const
     PendingTransactionImpl * transaction = new PendingTransactionImpl(*this);
 
     do {
-        if(!cryptonote::get_account_address_from_str(info, m_wallet->nettype(), dst_addr)) {
-            // TODO: copy-paste 'if treating as an address fails, try as url' from simplewallet.cpp:1982
-            setStatusError(tr("Invalid destination address"));
+        bool payment_id_seen = false;
+        bool payment_id_long_seen = false;
+        std::vector<uint8_t> extra;
+        vector<cryptonote::tx_destination_entry> dsts;
+        if (!amount && dst_addr.size() > 1) {
+            setStatusError(tr("Sending all requires one destination address"));
             break;
         }
-
-
-        std::vector<uint8_t> extra;
-        // if dst_addr is not an integrated address, parse payment_id
-        if (!info.has_payment_id && !payment_id.empty()) {
-            // copy-pasted from simplewallet.cpp:2212
-            crypto::hash payment_id_long;
-            bool r = tools::wallet2::parse_long_payment_id(payment_id, payment_id_long);
-            if (r) {
+        if (amount && (dst_addr.size() != (*amount).size())) {
+            setStatusError(tr("Destinations and amounts are unequal"));
+            break;
+        }
+        bool error = false;
+        for (size_t i = 0; i < dst_addr.size() && !error; i++) {
+            if(!cryptonote::get_account_address_from_str(info, m_wallet->nettype(), dst_addr[i])) {
+                // TODO: copy-paste 'if treating as an address fails, try as url' from simplewallet.cpp:1982
+                setStatusError(tr("Invalid destination address"));
+                error = true;
+                break;
+            }
+            if (info.has_payment_id || !payment_id.empty()) {
+                if (payment_id_seen || (info.has_payment_id && !payment_id.empty())) {
+                    setStatusError(tr("a single transaction cannot use more than one payment id"));
+                    error = true;
+                    break;
+                }
+                crypto::hash payment_id_long;
                 std::string extra_nonce;
-                cryptonote::set_payment_id_to_tx_extra_nonce(extra_nonce, payment_id_long);
-                r = add_extra_nonce_to_tx_extra(extra, extra_nonce);
-            } else {
-                r = tools::wallet2::parse_short_payment_id(payment_id, info.payment_id);
-                if (r) {
-                    std::string extra_nonce;
+                if (info.has_payment_id) {
                     set_encrypted_payment_id_to_tx_extra_nonce(extra_nonce, info.payment_id);
-                    r = add_extra_nonce_to_tx_extra(extra, extra_nonce);
+                    payment_id_seen = true;
+                } else if (tools::wallet2::parse_long_payment_id(payment_id, payment_id_long)) {
```

Comment:
I suggest to place long payment id logic outside the loop and set `payment_id_seen` flag there (if present). You won't need `payment_id_long_seen` flag then. And the condition will look like

```c++
            if (info.has_payment_id) {
                if (payment_id_seen) {
                    setStatusError(tr("a single transaction cannot use more than one payment id"));
                    error = true;
                    break;
                }
                ...
```

Reply:
Right that’s clearly better, updated.